### DISCUSSION TITLE
Sensor data - Modbus bridge tcp

### DISCUSSION THREAD

**Discussion by spueli99** [arendst/Tasmota#22716] [Q&A]
I would like to provide the data of a bosch bme680 under tasmota esp32 via modbus tcp for heating control. The control (technical alternative cmi) can only work as a Modbus master. It is unclear to me whether and how I can implement this using modbus bridge tcp.

[TA C.M.I. Modbus Input](https://help.ta.co.at/EN/CMIHELP/modbus_1.htm)

**Comment by barbudor** [comment]
None of the MODBUS implementation in Tasmota can make tasmota appear as a MODBUS device.
The MODBUS bridge is intended to bridge between MODBUS TCP protocol and MODBUS serial protocol to allow accessing MODBUS serial devices through TCP protocol

If you want to simulate a MODBUS TCP device on an ESP32, you would have to write you own. That is probably feasible in Tasmota using Berry

**Reply by mi-hol** [reply]
> None of the MODBUS implementation in Tasmota can make tasmota appear as a MODBUS device.

I wonder if this is a 100% correct statement, because [this script](https://github.com/CherAlban/tasmota-sml-modbus-gen24-domoticz-ha-script/blob/main/sml-modbus-mqtt.tas) makes an ESP32 pretend to be a Fronius Smart Meter and answers modbus requests

**Reply by sfromis** [reply]
As I read it, what you post cannot be called a feature of Tasmota, instead if seems to be someone writing scripter code to simulate Modbus TCP. While this looks useful, it is not part of the Tasmota project.

**Reply by mi-hol** [reply]
> what you post cannot be called a feature of Tasmota

well its build upon the supported tcpserver feature of Tasmota (commands below). 
Now instead of nit picking, lets see if OP is happy with the answer.
@spueli99 may I kindly ask for your view?

**TCP server support**
```
#define USE_SCRIPT_TCP_SERVER
wso(port) start a tcp stream server at port
wsc() close tcp stream server
wsa() return bytes available on tcp stream
wsrs() return a string read from tcp stream
wsws(string) writes a string to tcp stream
wsra(array) reads a tcp stream into array
wswa(array num (type)) writes num bytes of array to tcp stream, type: 0 = uint8 (default), 1 = uint16, 2 = sint16, 3 = float
```

**Reply by sfromis** [reply]
Exactly, what is part of Tasmota scripter is general functions for some TCP server, where you can "write your own" code for a specific protocol, like Modbus. Hence, the latter is not a feature of Tasmota.

I was merely commenting on your suggestion about the answer from barbudor not being a "correct statement", to ensure clarity about the details.

**Comment by Noschvie** [comment]
I think that a solution can be implemented with Node-Red to query the sensor data via Modbus TCP.

**Reply by mi-hol** [reply]
@spueli99 max I suggest to change the title to "Get sensor data via Modbus tcp" as this would spell out clearly what is desired and avoids confusion introduced by using "Modbus bridge tcp"
